fdb-extensions/src/main/java/com/apple/foundationdb/async/hnsw/HNSW.java

```diff
@@ -0,0 +1,2022 @@
+                                         @Nonnull final BiFunction<R, Node<N>, U> biMapFunction) {
+        final U bypass = fetchBypassFunction.apply(nodeReference);
+        if (bypass != null) {
+            return CompletableFuture.completedFuture(bypass);
+        }
+
+        return onReadListener.onAsyncRead(
+                        storageAdapter.fetchNode(readTransaction, storageTransform, layer,
+                                nodeReference.getPrimaryKey()))
+                .thenApply(node -> biMapFunction.apply(nodeReference, node));
+    }
+
+    /**
+     * Asynchronously fetches neighborhood nodes and returns them as {@link NodeReferenceWithVector} instances,
+     * which include the node's vector.
+     * <p>
+     * This method efficiently retrieves node data by first checking an in-memory {@code nodeCache}. If a node is not
+     * in the cache, it is fetched from the {@link StorageAdapter}. Fetched nodes are then added to the cache to
+     * optimize subsequent lookups. It also handles cases where the input {@code neighborReferences} may already
+     * contain {@link NodeReferenceWithVector} instances, avoiding redundant work.
+     *
+     * @param <N> the type of the node reference, extending {@link NodeReference}
+     * @param storageAdapter the storage adapter to fetch nodes from if they are not in the cache
+     * @param readTransaction the transaction context for database read operations
+     * @param storageTransform an affine transformation operator that is used to transform the fetched vector into the
+     *        storage space that is currently being used
+     * @param layer the graph layer from which to fetch the nodes
+     * @param neighborReferences an iterable of references to the neighbor nodes to be fetched
+     * @param nodeCache a map serving as an in-memory cache for nodes. This map will be populated with any
+     * nodes fetched from storage.
+     *
+     * @return a {@link CompletableFuture} that, upon completion, will contain a list of
+     * {@link NodeReferenceWithVector} objects for the specified neighbors
+     */
+    @Nonnull
+    private <N extends NodeReference> CompletableFuture<List<NodeReferenceWithVector>>
+            fetchNeighborhood(@Nonnull final StorageAdapter<N> storageAdapter,
+                              @Nonnull final ReadTransaction readTransaction,
+                              @Nonnull final AffineOperator storageTransform,
+                              final int layer,
+                              @Nonnull final Iterable<? extends NodeReference> neighborReferences,
+                              @Nonnull final Map<Tuple, Node<N>> nodeCache) {
+        return fetchSomeNodesAndApply(storageAdapter, readTransaction, storageTransform, layer, neighborReferences,
+                neighborReference -> {
+                    if (neighborReference instanceof NodeReferenceWithVector) {
+                        return (NodeReferenceWithVector)neighborReference;
```

Comment:
This is one of the places where I was thinking you'd replace the instance of check with:

```java
if (neighborReference.hasVector()) {
   return neighborReference.asNodeReferenceWithVector();
}
```

Reply:
Done.